### PR TITLE
[ROMM-55] Support for saves, states and screenshots

### LINKED ISSUE #55: [Feature] Savefiles/states management
**Is your feature request related to a problem? Please describe.**
Posibility to manage (download and upload) different savefiles/states for each game

### PR DESCRIPTION
**Pull request by gantoine**
### Supported internal folder structures

```
- library
  - roms
    - gbc
      - game_1.zip
  - saves
    - ps
      - PCSX-ReARMed
        - game.srm
    - snes
      - Snes9x 2005 Plus
        - game_1.srm
      - game_2.srm  # Can be stored outside emulator folders
  - states
    - gba
      - mGBA
        - game_1.state
      - game_2.state  # Can be stored outside emulator folders
  - screenshots
    - n64
      - game_1.png
    - game_2.png  # Can be stored outside platform folders
```

```
- library
  - gba
    - roms
      - game_1.zip
    - saves
      - mGBA
        - game_1.srm
      - game_2.srm  # Can be stored outside emulator folders
    - states
      - mGBA
        - game_1.state
      - game_2.state  # Can be stored outside emulator folders
    - screenshots
      - game_1.png
  - screenshots  # Not attributed to platform
    - game_2.png
```
Closes #55

### CODE REVIEW

File: backend/config/config_loader.py [hunk 1]
```diff
@@ -27,20 +31,46 @@ class ConfigDict(TypedDict):
...
+    EXCLUDED_SINGLE_FILES: list[str]
+    EXCLUDED_MULTI_FILES: list[str]
+    EXCLUDED_MULTI_PARTS_EXT: list[str]
+    EXCLUDED_MULTI_PARTS_FILES: list[str]
+    PLATFORMS_BINDING: dict[str, str]
+    ROMS_FOLDER_NAME: str
+    SAVES_FOLDER_NAME: str
+    STATES_FOLDER_NAME: str
+    SCREENSHOTS_FOLDER_NAME: str
+    HIGH_PRIO_STRUCTURE_PATH: str
```
Comment (gantoine) [inline backend/config/config_loader.py:53]: These need to be repeated twice, shame but necessary.

File: backend/endpoints/scan.py [hunk 2]
```diff
@@ -86,7 +104,106 @@ async def scan_platforms(
...
+            scanned_screenshot.platform_slug = scanned_platform.slug
+
+            rom = dbh.get_rom_by_filename_no_tags(scanned_screenshot.file_name_no_tags)
+            if rom:
+                scanned_screenshot.rom_id = rom.id
+                dbh.add_screenshot(scanned_screenshot)
+
+        dbh.purge_saves(scanned_platform.slug, [s for _e, s in fs_assets["saves"]])
+        dbh.purge_states(scanned_platform.slug, [s for _e, s in fs_assets["states"]])
+        dbh.purge_screenshots(fs_assets["screenshots"], scanned_platform.slug)
```
Comment (gantoine) [inline backend/endpoints/scan.py:197]: All of these are files nested under platforms, and for saves/states with optional emulator folders

File: backend/endpoints/scan.py [hunk 3]
```diff
@@ -86,7 +104,106 @@ async def scan_platforms(
...
+                screenshot.id, {"file_size_bytes": scanned_screenshot.file_size_bytes}
+            )
+            continue
+
+        rom = dbh.get_rom_by_filename_no_tags(scanned_screenshot.file_name_no_tags)
+        if rom:
+            scanned_screenshot.rom_id = rom.id
+            scanned_screenshot.platform_slug = rom.platform_slug
+            dbh.add_screenshot(scanned_screenshot)
+
```
Comment (gantoine) [inline backend/endpoints/scan.py:224]: This is for screenshots manually added, or where the user might not place screenshots under a specific platform slug

File: backend/models/assets.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+    rom_id = Column(
+        Integer(), ForeignKey("roms.id", ondelete="CASCADE"), nullable=False
+    )
+    rom = relationship("Rom", lazy="selectin", back_populates="screenshots")
+
+    platform_slug = Column(
+        String(length=50),
+        ForeignKey("platforms.slug", ondelete="CASCADE"),
+        nullable=True,
```
Comment (gantoine) [inline backend/models/assets.py:86]: Optional since it might not be nested under a platform folder

File: backend/models/assets.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    @cached_property
+    def full_path(self) -> str:
+        return f"{self.file_path}/{self.file_name}"
+
+    @cached_property
+    def download_path(self) -> str:
+        return f"{FRONTEND_LIBRARY_PATH}/{self.full_path}"
+
+
+class Save(BaseAsset):
```
Comment (gantoine) [inline backend/models/assets.py:36]: All assets are required to match a rom before being ingested by the db.

File: backend/models/rom.py [hunk 6]
```diff
@@ -88,6 +107,12 @@ def has_cover(self) -> bool:
             or self.path_cover_l != DEFAULT_PATH_COVER_L
         )
 
+    @cached_property
+    def merged_screenshots(self) -> list[str]:
```
Comment (gantoine) [inline backend/models/rom.py:112]: sourced from igdb and uploaded screenshots

File: backend/utils/fastapi.py [hunk 7]
```diff
@@ -113,19 +114,66 @@ async def scan_rom(
...
+    )
+
+    #  Scan asset with the sames path and emulator folder name
+    if emulator:
+        return State(**_scan_asset(file_name, os.path.join(states_path, emulator)))
+
+    return State(**_scan_asset(file_name, states_path))
+
+
+def scan_screenshot(file_name: str, fs_platform: str = None) -> Screenshot:
```
Comment (gantoine) [inline backend/utils/fastapi.py:171]: This function supports scanning under a platform folder and generically

File: backend/alembic/versions/0014_asset_files.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+Revises: 0013_upgrade_file_extension
+Create Date: 2024-01-07 09:45:09.632571
+
+"""
+from alembic import op
+import sqlalchemy as sa
+
+
+# revision identifiers, used by Alembic.
+revision = 'ca41fa04d88c'
```
Comment (zurdi15) [inline backend/alembic/versions/0014_asset_files.py]: This may be a misslick?

**Comment by zurdi15** [comment]
Also i'm testing to upload user ``avatar`` and it's failing, passing a ``undefined`` to the backend, but the ``custom cover art`` works perfectly fine.

**Comment by gantoine** [comment]
> Also i'm testing to upload user avatar and it's failing

Good catch, fixed in the last commit. Seems it broke during the typescript rewrite.

**Comment by zurdi15** [comment]
Just saw this, are we going to do it? https://github.com/zurdi15/romm/issues/55#issuecomment-1532770116

**Comment by gantoine** [comment]
> Just saw this, are we going to do it? The renaming thing

I forgot about that, easy enough to do so I'm going to add it in this PR

File: backend/config/config_loader.py (hunk 1 above)
Comment (zurdi15) [reply backend/config/config_loader.py:53]: why need to be repeated twice?

Comment (gantoine) [reply backend/config/config_loader.py:53]: there's no way to use the typedef in the class, or the other way around. even asked chatgpt and it had no clue lmao.

**Review by zurdi15 — DISMISSED** [review]

**Review by zurdi15 — APPROVED** [review]

**Pull request merged**